src/selectedtests/test_mappings/mappings.py

```diff
@@ -0,0 +1,125 @@
+        self._file_count_map = file_count_map
+        self._project = project
+        self._repo_name = repo_name
+        self._branch = branch
+        self._test_mappings = None
+
+    @classmethod
+    def create_mappings(
+        cls,
+        repo: Repo,
+        source_re: Pattern,
+        test_re: Pattern,
+        start_date: datetime,
+        end_date: datetime,
+        project: str,
+        branch: str,
+    ):
+        """
+        Create the test mappings for an evergreen project. Optionally looks at an associated module.
+
+        :param repo: The repo that contains the source code for the evergreen project.
+        :param source_re: Regex pattern to match changed source files against.
+        :param test_re: Regex pattern to match changed test files against.
+        :param start_date: The date at which to start analyzing commits of the project.
+        :param end_date: The date up to which we should analyze commits of the project.
+        :param project: The name of the evergreen project to analyze.
+        :param branch: The branch of the git repo used for the evergreen project.
+        :return: An instance of the test mappings class
+        """
+        file_intersection = defaultdict(lambda: defaultdict(int))
+        file_count = defaultdict(int)
+
+        LOGGER.debug("searching from", ts=start_date)
+        LOGGER.debug("searching until", ts=end_date)
+        for commit in repo.iter_commits(repo.head.commit):
+            LOGGER.debug(
+                "Investigating commit",
+                summary=commit.message.splitlines()[0],
+                ts=commit.committed_datetime,
+                id=commit.hexsha,
+            )
+
+            if commit.committed_datetime.timestamp() < start_date.timestamp():
+                break
+
+            if commit.committed_datetime.timestamp() > end_date.timestamp():
+                continue
+
+            tests_changed = set()
+            src_changed = set()
+            for path in modified_files_for_commit(commit, LOGGER):
+                LOGGER.debug("found change", path=path)
+
+                if test_re.match(path):
+                    tests_changed.add(path)
+                elif source_re.match(path):
+                    src_changed.add(path)
+
+            for src in src_changed:
+                file_count[src] += 1
+                for test in tests_changed:
+                    file_intersection[src][test] += 1
+
+        repo_name = os.path.basename(repo.working_dir)
+        return TestMappings(file_intersection, file_count, project, repo_name, branch)
+
+    def get_mappings(self):
+        """
+        Get a transformed version of test mappings to the test mapping object.
+
+        :return: Transformed test mappings
+        """
+        if not self._test_mappings:
+            self._test_mappings = self._transform_mappings()
+        return self._test_mappings
+
+    def _transform_mappings(self):
```

Comment:
Can you add some tests for this? To make sure that the format that you expect will be what comes back

Reply:
That is implicitly tested in the `get_mappings` tests

I tend to favor testing private methods through the public methods that call them: https://www.codewithjason.com/testing-private-methods/